File: Algorithms/Searching & Sorting/QuickSort/quicksort.java
class QuickSort { 
    static int partition(int elem[], int low, int high) { 
        int pivot = elem[high];  
        int i = (low-1); // index of smaller element 
        for (int j=low; j<high; j++)
            if (elem[j] < pivot) { // check scale for elements less than pivot
                i++; 
                int temp = elem[i]; 
                elem[i] = elem[j]; 
                elem[j] = temp;  // swap
            } 

        int temp = elem[i+1]; 
        elem[i+1] = elem[high]; 
        elem[high] = temp; // swap
        return i+1; 
    } 
  
    static void quickSort(int elem[], int low, int high) { 
        if (low < high) { 
            int pi = partition(elem, low, high); 
            quickSort(elem, low, pi-1); 
            quickSort(elem, pi+1, high); 
        } 
    } 
  
    // result_print
    static void print_result(int elem[]) {
        int n = elem.length; 
        for (int i=0; i<n; ++i) 
            System.out.print(elem[i]+" "); 
    } 
  
    // Driver function
    public static void main(String args[]) { 
        int elem[] = {9, 7, 2, 3, 7, 11, 01, 90, 12, 82}; 
        int n = elem.length; 
  
        QuickSort qsj = new QuickSort(); 
        qsj.quickSort(elem, 0, n-1);

        System.out.println("The sorted array is:"); 
        print_result(elem); 
    } 
}

File: Algorithms/Searching & Sorting/QuickSort/quickSort.cpp
#include<iostream>
using namespace std;

// Partition
int partition(int *array, int l, int r){
    int pivot = array[r];
    int i = l-1; // place for swapping
    for(int j = l; j <= r-1; j++){
        if(array[j] <= pivot){
            i++;
            swap(array[i], array[j]);
        }
    }
    swap(array[i+1], array[r]);
    return (i+1);
}

// Random pivot
int partition_random_pivot(int *array, int l, int r){
    srand(time(NULL));
    int random = l + rand()%(r - l);
    swap(array[random], array[r]);
    return partition(array, l, r);
}

int quick_sort(int *array, int l, int r){
    if(l < r){   // base case
        int pi = partition_random_pivot(array, l, r);  // returns index of pivot
        quick_sort(array, l, pi-1);
        quick_sort(array, pi+1, r);
    }
}

int main(int argc, char const *argv[]){
	int N;
	cout << "Enter the size of array:";
	cin >> N;
	int A[N];
	cout << "Enter the array elements:";
	for(int i = 0; i < N; i++) cin >> A[i];
	
	quick_sort(A, 0, N-1);
	cout << "Sorted array is: ";
	for(int i = 0; i < N; i++) cout << A[i] << " ";
	cout << endl;

	return 0;
}
